When registering patterns include postTypes

dsas opened the pull request:
#### Proposed Changes

When registering block patterns from APIs set the [`postTypes`](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/#register_block_pattern) pattern property to `wp_template` and `wp_template_part` for patterns that have a block_type of `core/template-part/*`.

The practical effect of this right now is that header and footer patterns from our pattern repositories will only be offered inside the site editor and not inside the post or page editor.

The downside is that header and footer categories will still exist, as core and themes generally don't seem to have enacted this change yet. I think this may make this change confusing with little gain, but we need to break the chicken/egg cycle somewhere


#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to wordpress.com and create a new post
* Open the pattern selector (big plus top left)
* Change category to footers and see a big list including patterns like 'Simple centered footer'
* Apply the update using the instructions below
* You may also need to add `define('WP_DISABLE_PATTERN_CACHE',1);` to your 0-sandbox.php
* Refresh the page and open up the pattern selector again
* Note that there are fewer patterns under headers and footers. All of the dotcompatterns no longer appear including 'Simple centered footer'
* Open up the site-editor
* Open up the patterns, see that the full pattern list is in headers and footers again, including everything from dotcompatterns and  'Simple centered footer'
* Note that this only affect headers and footers

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #66628

dsas committed:
When registering patterns include postTypes

When registering block patterns from APIs set the `postTypes` pattern
property to `wp_template` and `wp_template_part` for patterns that have
a block_type of `core/template-part/*`.

The practical effect of this right now is that header and footer
patterns from our pattern repositories will only be offered inside the
site editor and not inside the post or page editor

Fixes #66628

dsas committed:
Add some unit tests

dsas committed:
Fix tests

dsas commented:
Neither Core or 2023 use postTypes header and footers yet. I've gone back to p1668521835785829/1660679319.921149-slack-C02M88KJ684 to confirm we're ok to merge this regardless, and whether I can help core/2023 adopt postTypes.

dsas commented:
Filed [57135](https://core.trac.wordpress.org/ticket/57135) for TT3 and https://github.com/WordPress/gutenberg/issues/45858 for Gutenberg bundled patterns. It'll be interesting to see what the plans are in core.

dsas commented:
This PR uses Gutenberg pattern registration APIs so that header and footer patterns from pattern source sites are only displayed when editing templates and not when editing pages & posts. For context, the initial request was https://github.com/Automattic/wp-calypso/issues/66628

Patterns registered from other sources (active theme, gutenberg bundles) are not forced to register patterns in the same way though, and don't. I opened issues to try and find out what is happening in [Twenty Twenty-Three](https://core.trac.wordpress.org/ticket/57135) (no response) and in [Gutenberg itself](https://github.com/WordPress/gutenberg/issues/45858) (one response saying it may make sense in some contexts)

I've held off merging this PR because I'm concerned that without achieving some consensus on this, it's create a surprising experience - a user would see that "random" patterns don't appear in the post editor, but do in the site editor.

I think there are a few ways forward:
 1. Amplify the Core/GB discussions and abide by any consensus arrived at. However 3rd party themes may never abide by that consensus.
 2. Write some code to unregister all patterns in the header/footer category and then reregister them, setting them to only be available when editing templates. 
 3. Drop this PR and don't follow up on the idea.

I think the way forward is the second option. I think this should be technically straight forward. We alread unregister and reregister some core patterns to fit in our categorisation structure.

ianstewart commented:
> Write some code to unregister all patterns in the header/footer category and then reregister them, setting them to only be available when editing templates.

Would this apply to third-party themes on Dotcom as well? Not a blocker either way, just wanted to check.

> We probably should follow the https://github.com/WordPress/gutenberg/issues/45858#issuecomment-1318797015 and allow headers/footers on pages but not posts

This idea makes sense to me as well.

dsas committed:
More comprehensively add postTypes to patterns

As well as patterns from the source site, also add postTypes to any
other registered patterns that are intended for template parts.

Also decided to allow pages to use template-part patterns, as people
often use a blank template and add in more template-y patterns for
one-off pages.

dsas commented:
> Would this apply to third-party themes on Dotcom as well?

The same logic will be applied to all themes. So yes, so long as they've declared that the pattern is intended for blockTypes that are templates. Our block themes usually do, TT3 included. 

I tried to sample some third-party .org themes and they mainly don't....but then patterns with a third party theme can be a bit too many cooks spoil the broth
![image](https://user-images.githubusercontent.com/93301/209352992-75c69c1c-7ef9-4682-b87a-c32df3f28572.png)

dsas committed:
Fix tests, hopefully

dsas committed:
Fix tests

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns-from-api.php:
```diff
@@ -117,27 +117,35 @@ function ( $a, $b ) {
+						'blockTypes'    => $block_types,
+					);
+
+					$post_types = $this->utils->get_pattern_post_types_from_pattern( $pattern_props );
+					if ( $post_types ) {
+						$pattern_props['postTypes'] = $post_types;
+					}
```
Nitpick: Since we're running `update_pattern_post_types` now on all registered patterns, I think we can drop this piece of code.

dsas committed:
Remove duplicate pattern post types call

mmtr reviewed (approved):
Works great! I confirmed the patterns remain the same for both pages and templates.
